subsys/net/lib/sockets/sockets.c

```diff
@@ -980,8 +980,13 @@ int zsock_wait_data(struct net_context *ctx, k_timeout_t *timeout)
 
 	if (k_fifo_is_empty(&ctx->recv_q)) {
 		/* Wait for the data to arrive but without holding a lock */
-		return k_condvar_wait(&ctx->cond.recv, ctx->cond.lock,
+		int ret = k_condvar_wait(&ctx->cond.recv, ctx->cond.lock,
 				      *timeout);
+		/* Release the lock because the condition variable is
+		 * just used for signaling purpose.
+		 */
+		k_mutex_unlock(ctx->cond.lock);
```

Comment:
I don't think it's right - one of the purpose of `k_condvar_wait()` is to restore the lock once it quits - the socket lock will be released in the corresponding `recv()`/`recvfrom()` call.

Reply:
You are right that the lock is aquiered when calling `recv()`/`recvfrom()` but when calling `read()` it enters the `k_condvar_wait()` without previously owning the lock so that's where the uneven number of lock/unlock is happening.

Is this behavior corret? Should I only call `recv()` and not `read()` from my app?

Reply:
Ok, that's not right then, `read()` should also acquire the mutex. It seems that this should be addressed in the `fdtable.c` which provides generic `read()` implementation (I think the mutex should be acquired before vtable call, just like in case other calls).

Reply:
These changes in PR works well for me. I dont't know if I should add the locking to other functions working with vtable like `write()`, `close()`?

Reply:
I think it'd make sense, both for consistency and additional protection.